File: fpga/fomu/csm/ucode.js
// ucode.js -- uCode (Forth dialect) compiler
// Dale Schumacher
// 2024-04-10

function error(...msg) {
//debug console.log("ERROR!", ...msg);
    return msg.join(" ");
}

function compile(text) {
    let pos = 0;  // input position within `text`
    function next_char() {
        const cp = text.codePointAt(pos);  // returns `undefined` if out-of-bounds
        if (typeof cp === "number") {
            pos += (cp <= 0xFFFF ? 1 : 2);
        }
        return cp;
    }
    function next_token() {
        let token = "";
        // skip whitespace
        let cp = next_char();
        while ((typeof cp === "number") && (cp <= 32)) {
            cp = next_char();
        }
        // collect non-whitespace
        while ((typeof cp === "number") && (cp > 32)) {
            token += String.fromCodePoint(cp);
            cp = next_char();
        }
        // return completed token
        return token;
    }

    const ADDR_MASK =           0x0FFF;                 // 12-bit uCode addresses

    function uc_call(addr) {    // push return address and jump
        return (0xC000 | (addr & ADDR_MASK));
    }
    function uc_jump(addr) {    // jump (unconditional)
        return (0x8000 | (addr & ADDR_MASK));
    }
    function uc_jz(addr) {      // jump if zero
        return (0xA000 | (addr & ADDR_MASK));
    }
    function uc_jz_inc(addr) {  // jump if zero, else increment
        return (0x9000 | (addr & ADDR_MASK));
    }
    function uc_jz_dec(addr) {  // jump if zero, else decrement
        return (0xB000 | (addr & ADDR_MASK));
    }

    const UC_LIT =          0x021F;                     // (LIT) item ( -- item )
    const UC_CONST =        0x521F;                     // (CONST) item ( -- item ) ( R: addr -- ) addr->pc
    const UC_TO_R =         0x2100;                     // >R ( a -- ) ( R: -- a )
    const UC_R_FROM =       0x1280;                     // R> ( -- a ) ( R: a -- )
    const UC_R_FETCH =      0x0280;                     // R@ ( -- a ) ( R: a -- a )
    const UC_EXIT =         0x5000;                     // EXIT ( -- ) ( R: addr -- ) addr->pc

    const words = {
        "NOP":              0x0000,                     // ( -- )
        "DROP":             0x0100,                     // ( a -- )
        "DUP":              0x0200,                     // ( a -- a a )
        "SWAP":             0x0400,                     // ( a b -- b a )
        "OVER":             0x0240,                     // ( a b -- a b a )
        "ROT":              0x0500,                     // ( a b c -- b c a )
        "-ROT":             0x0600,                     // ( a b c -- c a b )
        "TRUE":             0x02F6,                     // ( -- -1 )
        "FALSE":            0x02C0,                     // ( -- 0 )
        "0":                0x02C0,                     // ( -- 0 )
        "1":                0x02D6,                     // ( -- 1 )
        "-1":               0x02F6,                     // ( -- -1 )
        "LSB":              0x02D6,                     // ( -- 1 )
        "MSB":              0x02E6,                     // ( -- -32768 )
        "LSB&":             0x0314,                     // ( a -- a&1 )
        "MSB&":             0x0324,                     // ( a -- a&-32768 )
        "INVERT":           0x0335,                     // ( a -- ~a )
        "NEGATE":           0x03C2,                     // ( a -- -a )
        "1+":               0x0311,                     // ( a -- a+1 )
        "1-":               0x0312,                     // ( a -- a-1 )
        "2*":               0x0301,                     // ( a -- a*2 )
        "2/":               0x030B,                     // ( a -- a/2 )
        "+":                0x0741,                     // ( a b -- a+b )
        "-":                0x0742,                     // ( a b -- a-b )
        "*":                0x0743,                     // ( a b -- a*b )
        "AND":              0x0744,                     // ( a b -- a&b )
        "XOR":              0x0745,                     // ( a b -- a^b )
        "OR":               0x0746,                     // ( a b -- a|b )
        "ROL":              0x0307,                     // ( a -- {a[14:0],a[15]} )
        "2ROL":             0x0308,                     // ( a -- {a[13:0],a[15:14]} )
        "4ROL":             0x0309,                     // ( a -- {a[11:0],a[15:12]} )
        "8ROL":             0x030A,                     // ( a -- {a[7:0],a[15:8]} )
        "ASR":              0x030B,                     // ( a -- {a[15],a[15:1]} )
        "2ASR":             0x030C,                     // ( a -- {a[15],a[15],a[15:2]} )
        "4ASR":             0x030D,                     // ( a -- {a[15],a[15],a[15],a[15],a[15:4]} )
        "@":                0x030F,                     // ( addr -- data )
        "!":                0x098F,                     // ( data addr -- )
        "IO@":              0x033F,                     // ( io_reg -- data )
        "IO!":              0x09BF,                     // ( data io_reg -- )
        "T@":               0x034F,                     // ( qref -- data )
        "T!":               0x09CF,                     // ( data qref -- )
        "X@":               0x035F,                     // ( qref -- data )
        "X!":               0x09DF,                     // ( data qref -- )
        "Y@":               0x036F,                     // ( qref -- data )
        "Y!":               0x09EF,                     // ( data qref -- )
        "Z@":               0x037F,                     // ( qref -- data )
        "Z!":               0x09FF,                     // ( data qref -- )
        ">R":               0x2100,                     // ( a -- ) ( R: -- a )
        "R>":               0x1280,                     // ( -- a ) ( R: a -- )
        "R@":               0x0280,                     // ( -- a ) ( R: a -- a )
        "RDROP":            0x1000,                     // ( -- ) ( R: a -- )
        "EXIT":             0x5000                      // ( -- ) ( R: addr -- ) addr->pc ; no TCO
    };
    words[":"] = function () {
        // new entry-point
        const word = uc_call(prog.length);
        prog[0] = word;  // update bootstrap entry-point
        const name = next_token();
//debug console.log("compile_name:", name, "=", word.toString(16).padStart(4, "0"));
        words[name] = word;  // add word to dictionary
    };
    words["VARIABLE"] = function () {
        // new named variable
        const word = uc_call(prog.length);
        const name = next_token();
//debug console.log("compile_var:", name, "=", word.toString(16).padStart(4, "0"));
        prog.push(UC_CONST);
        prog.push(prog.length + 1);  // variable address
        prog.push(0);  // variable data field
        words[name] = word;  // add word to dictionary
    };
    words["SKZ"] = function () {                        // ( 0 -- ) pc+2->pc | ( n -- )
        // skip (next instruction), if TOS is zero
        prog.push(uc_jz(prog.length + 2));
    };
    words["BEGIN"] = function () {                      // ( -- )
        // begin indefinite loop
        const addr = prog.length;
//debug console.log("compile_indefinite_loop:", "$"+addr.toString(16).padStart(3, "0"));
        const word = uc_jump(addr);  // placeholder
        ctrl_ctx.push(word);
    };
    words["UNTIL"] = function () {                      // ( cond -- )
        // end bottom-test loop
        const addr = ctrl_ctx.pop() & ADDR_MASK;
        prog.push(uc_jz(addr));
    };
    words["WHILE"] = function () {                      // ( cond -- )
        // loop (top) test
        const addr = ctrl_ctx.pop() & ADDR_MASK;
        const word = uc_jump(prog.length);
        ctrl_ctx.push(word);
        prog.push(uc_jz(addr));  // placeholder
    };
    words["REPEAT"] = function () {                     // ( -- )
        // end top-test loop
        const addr = ctrl_ctx.pop() & ADDR_MASK;
        const word = uc_jump(prog[addr]);
        prog.push(word);
        prog[addr] = uc_jz(prog.length);  // patch
    };
    words["?D0"] = function () {                        // ( n -- ) ( R: -- n' )
        // begin counted loop
        const addr = prog.length;
//debug console.log("compile_counted_loop:", "$"+addr.toString(16).padStart(3, "0"));
        const word = uc_jz(addr);  // placeholder
        ctrl_ctx.push(word);
        prog.push(word);
        prog.push(UC_TO_R);
    };
    words["I"] = function () {                          // ( -- n ) ( R: n -- n )
        // fetch loop count (from R-stack)
        prog.push(UC_R_FETCH);
    };
    words["LOOP-"] = function () {                      // loop->pc
        // end decrement loop
        prog.push(UC_R_FROM);
        const addr = ctrl_ctx.pop() & ADDR_MASK;
        prog.push(uc_jump(addr));
        prog[addr] = uc_jz_dec(prog.length);  // patch
    };
    words["LOOP+"] = function () {                      // loop->pc
        // end increment loop
        prog.push(UC_R_FROM);
        const addr = ctrl_ctx.pop() & ADDR_MASK;
        prog.push(uc_jump(addr));
        prog[addr] = uc_jz_inc(prog.length);  // patch
    };
    words["IF"] = function () {                         // ( cond -- )
        // begin conditional
//debug console.log("compile_if:", "$"+prog.length.toString(16).padStart(3, "0"));
        const word = uc_jz(prog.length);  // placeholder
        ctrl_ctx.push(word);
        prog.push(word);
    };
    words["ELSE"] = function () {                       // ( -- )
        // begin alternative
//debug console.log("compile_else:", "$"+prog.length.toString(16).padStart(3, "0"));
        const addr = ctrl_ctx.pop() & ADDR_MASK;
        prog[addr] = uc_jz(prog.length + 1);  // patch
        const word = uc_jump(prog.length);  // placeholder
        ctrl_ctx.push(word);
        prog.push(word);
    };
    words["THEN"] = function () {                       // ( -- )
        // end conditional
//debug console.log("compile_then:", "$"+prog.length.toString(16).padStart(3, "0"));
        const word = ctrl_ctx.pop();
        const addr = word & ADDR_MASK;
        prog[addr] = (word & ~ADDR_MASK) | (prog.length & ADDR_MASK);  // patch
    };

    const ctrl_ctx = [];  // stack of flow-control contexts

    const prog = [
        uc_jump(0)
    ];
    function compile_words(token) {
        const TAIL_NONE = 0;
        const TAIL_EVAL = 1;
        const TAIL_CALL = 2;
        const TAIL_DATA = 3;
        let tail_ctx = TAIL_NONE;
        while (token.length > 0) {
//debug console.log("compile_words:", token);
            if (token === "(") {
                compile_comment(next_token());
            } else if (token === ";") {
                if (ctrl_ctx.length > 0) {
                    return error("EXIT from control depth", ctrl_ctx.length);
                }
                if (tail_ctx === TAIL_EVAL) {
                    // attach "free" EXIT to previous word
                    prog[prog.length - 1] |= UC_EXIT;
                } else if (tail_ctx === TAIL_CALL) {
                    // convert previous CALL to JUMP
                    prog[prog.length - 1] &= ~0x4000;
                } else {
                    // compile EXIT
                    prog.push(UC_EXIT);
                }
                tail_ctx = TAIL_NONE;
            } else if (token === "CONSTANT") {
                // allocate constant word
                const name = next_token();
                if (tail_ctx !== TAIL_DATA) {
                    return error("invalid constant:", name);
                }
                const addr = prog.length - 2;
                const word = uc_call(addr);
//debug console.log("compile_const:", name, "=", word.toString(16).padStart(4, "0"));
                prog[addr] = UC_CONST;  // convert (LIT) to (CONST)
                words[name] = word;  // add word to dictionary
                tail_ctx = TAIL_NONE;
            } else {
                const word = words[token];
                if (typeof word === "number") {
                    // compile primitive or call
                    prog.push(word);
                    tail_ctx = (word & 0xF000) === 0xC000
                        ? TAIL_CALL
                        : (word & 0xF000) === 0x0000
                            ? TAIL_EVAL
                            : TAIL_NONE;
                } else if (typeof word === "function") {
                    // invoke compiler function
                    word();
                    tail_ctx = TAIL_NONE;
                } else {
                    const num = Number(token);
                    if (Number.isSafeInteger(num)) {
                        // push number literal
                        prog.push(UC_LIT);
                        prog.push(num & 0xFFFF);  // truncate to 16 bits
                        tail_ctx = TAIL_DATA;
                    } else {
                        return error("invalid token:", token);
                    }
                }
            }
            token = next_token();
        }
        return token;
    }
    function compile_comment(token) {
        while (token.length > 0) {
//debug console.log("compile_comment:", token);
            if (token === "(") {
                compile_comment(next_token());
            } else if (token == ")") {
                break;
            }
            token = next_token();
        }
        return token;
    }

    const end = compile_words(next_token());
    if (end.length !== 0) {
        return end;
    }
    return prog;
}

//debug const simple_source = ": BOOT R@ DROP BOOT ;";
//debug const multiline_source = `
//debug 0x0FFF CONSTANT ADDR_MASK
//debug VARIABLE COUNTER
//debug 
//debug : ADJUST ( n -- n+COUNTER )
//debug     COUNTER @ +
//debug     DUP COUNTER ! ;
//debug : EXECUTE ( addr -- ) ( R: -- addr )
//debug     ADDR_MASK ( 0x0FFF ) AND >R
//debug : (EXIT)
//debug     EXIT
//debug : NIP ( a b -- b )
//debug     SWAP DROP ;
//debug : TUCK ( a b -- b a b )
//debug     SWAP OVER ;
//debug : ?: ( altn cnsq cond -- cnqs | altn )
//debug     SKZ SWAP
//debug : (DROP)
//debug     DROP ;
//debug : 0= ( n -- n==0 )
//debug : NOT ( flag -- !flag )
//debug     TRUE FALSE ROT ?: ;
//debug : BOOL ( n -- flag )
//debug     IF TRUE ELSE FALSE THEN ;
//debug : 0< ( n -- n<0 )
//debug     MSB& BOOL ;
//debug : 4DROP ( a b c d -- )
//debug     4 ?D0 DROP I DROP LOOP- ;
//debug : EMIT ( ch -- )
//debug     BEGIN 0x00 IO@ UNTIL 0x01 IO! ;
//debug : KEY ( -- ch )
//debug     BEGIN 0x02 IO@ NOT WHILE REPEAT 0x03 IO@ ;
//debug 
//debug ( WARNING! BOOT should not return... )
//debug : BOOT
//debug     R> DROP BOOT`;
// const source = simple_source;
//debug const source = multiline_source;
// console.log(compile(source));
//debug console.log(compile(source).map(function (number, index) {
//debug    return index.toString(16).padStart(3, "0") + ": " + number.toString(16).padStart(4, "0");
//debug }).join("\n"));

export default Object.freeze(compile);

File: fpga/fomu/csm/ucode.f
(
    Base word dictionary for uCode
)

: PANIC! FAIL PANIC! ;      ( if BOOT returns... )

0x03 CONSTANT ^C
0x08 CONSTANT '\b'
0x09 CONSTANT '\t'
0x0A CONSTANT '\n'
0x0D CONSTANT '\r'
0x20 CONSTANT BL
0x30 CONSTANT '0'
0x39 CONSTANT '9'
0x37 CONSTANT A-10
0x41 CONSTANT 'A'
0x46 CONSTANT 'F'
0x5A CONSTANT 'Z'
0x61 CONSTANT 'a'
0x66 CONSTANT 'f'
0x7A CONSTANT 'z'
0x7F CONSTANT DEL

: (JMP)
    R>
: @EXECUTE
    @
: EXECUTE
    0x0FFF AND >R
: (EXIT)
    EXIT
: ?: ( altn cnsq cond -- cnqs | altn )
    SKZ SWAP
: (DROP)
    DROP ;
: (VAR)
    R> ;
: (CONST)
    R> @ ;
: NIP ( a b -- b )
    SWAP DROP ;
: TUCK ( a b -- b a b )
    SWAP OVER ;
: 2DUP ( a b -- a b a b )
    OVER OVER ;
: 2DROP ( a b -- )
    DROP DROP ;

: ABS ( n -- +n )
    DUP MSB& SKZ NEGATE ;
: BOOL ( n -- flag )
    IF
        TRUE
    ELSE
        FALSE
    THEN ;
: = ( a b -- a==b )
    XOR
: 0= ( n -- n==0 )
: NOT ( flag -- !flag )
    IF
        FALSE
    ELSE
        TRUE
    THEN ;
: <> ( a b -- a!=b )
    = INVERT ;
: 0> ( n -- n>0 )
    NEGATE
: 0< ( n -- n<0 )
    MSB& BOOL ;
: > ( a b -- a>b )
    SWAP
: < ( a b -- a<b )
    - 0< ;
: >= ( a b -- a>=b )
    < INVERT ;
: <= ( a b -- a<=b )
    > INVERT ;
: MAX ( a b -- a | b )
    2DUP < ?: ;
: MIN ( a b -- a | b )
    2DUP > ?: ;
: INBOUNDS ( n lo hi -- lo<=n&&n<=hi )
    -ROT OVER SWAP -
    -ROT - OR MSB& 0= ;
: ISDIGIT ( char -- flag )
    '0' '9' INBOUNDS ;
: ISHEX ( char -- flag )
    DUP 'A' 'F' INBOUNDS
    OVER 'a' 'f' INBOUNDS OR
    SWAP ISDIGIT OR ;
: ISUPPER ( char -- flag )
    'A' 'Z' INBOUNDS ;
: ISLOWER ( char -- flag )
    'a' 'z' INBOUNDS ;
: TOUPPER ( 'A' | 'a' -- 'A' )
    DUP ISLOWER IF
        BL XOR  ( flip case )
    THEN ;
: TOLOWER ( 'A' | 'a' -- 'a' )
    DUP ISUPPER IF
        BL XOR  ( flip case )
    THEN ;

: TX? ( -- ready )
: EMIT?
    0x00 IO@ ;
: TX! ( char -- )
    0x01 IO! ;
: RX? ( -- ready )
: KEY?
    0x02 IO@ ;
: RX@ ( -- char )
    0x03 IO@ ;

: EMIT ( char -- )
    BEGIN TX? UNTIL TX! ;
: KEY ( -- char )
    BEGIN RX? UNTIL RX@ ;
: SPACE ( -- )
    BL EMIT ;
: SPACES ( n -- )
    ?D0
        SPACE
    LOOP- ;
: CR ( -- )
    '\r' EMIT '\n' EMIT ;
: >X ( n -- x )
    0xF AND
    DUP 10 < IF
        '0'
    ELSE
        A-10
    THEN + ;
: X> ( x -- n )
    TOUPPER
    DUP 'A' < IF
        '0'
    ELSE
        A-10
    THEN - ;
: X# ( n -- )
    >X EMIT ;
: X. ( n -- )
    4 ?D0
        4ROL DUP X#
    LOOP- DROP ;

( Debugging Monitor )
0x21 CONSTANT '!'
0x2E CONSTANT '.'
0x2F CONSTANT '/'
0x3C CONSTANT '<'
0x3E CONSTANT '>'
0x3F CONSTANT '?'
0x40 CONSTANT '@'
0x5B CONSTANT '['
0x5D CONSTANT ']'
0x71 CONSTANT 'q'
0x72 CONSTANT 'r'
VARIABLE cmd    ( last command character read )
VARIABLE inp    ( input data accumulator )
VARIABLE tos    ( top of stack )
VARIABLE nos    ( next on stack )
VARIABLE copy   ( bulk copy mode )
VARIABLE here   ( bulk copy addr )
: @1+ ( addr -- )
    DUP @ 1+ SWAP ! ;
: push ( a -- )
    tos @ nos !
    tos ! ;
: pop ( -- a )
    tos @
    nos @ tos ! ;
: quad ( raw -- addr )
    DUP MSB& IF
        0x0FFF AND          ( uCode | fixnum )
    ELSE
        2ROL
        DUP LSB& IF         ( RAM address )
            0x3FFC AND 0x4000 OR
        ELSE                ( ROM address )
            0x7FFC AND MSB|
        THEN
    THEN ;
: fetch ( addr -- data )
    DUP 0xC000 AND IF
        DUP 0x3 AND SWAP    ( D: field addr )
        DUP MSB& IF
            2ASR 0x1FFF
        ELSE
            2ASR 0x0FFF
        THEN AND            ( D: field offset )
        OVER 0x1 = IF
            QX@             ( D: field data )
        ELSE
            OVER 0x2 = IF
                QY@         ( D: field data )
            ELSE
                OVER 0x3 = IF
                    QZ@     ( D: field data )
                ELSE
                    QT@     ( D: field data )
                THEN
            THEN
        THEN
        NIP                 ( D: data )
    ELSE
        @                   ( D: data )
    THEN ;
: store ( data addr -- )
    DUP 0xC000 AND IF
        DUP 0x3 AND SWAP    ( D: data field addr )
        DUP MSB& IF
            2ASR 0x1FFF
        ELSE
            2ASR 0x0FFF
        THEN AND SWAP       ( D: data offset field )
        DUP 0x1 = IF
            DROP QX!
        ELSE
            DUP 0x2 = IF
                DROP QY!
            ELSE
                DUP 0x3 = IF
                    DROP QZ!
                ELSE
                    DROP QT!
                THEN
            THEN
        THEN
    ELSE
        !
    THEN ;
: dump ( start end -- )
    OVER -                  ( D: start span )
    DUP 0< IF
        2DROP
    ELSE
        1+ ?D0
            DUP fetch       ( D: addr data )
            OVER 0x7 AND IF
                SPACE
            ELSE
                CR
            THEN
            X. 1+           ( D: addr+1 )
        LOOP- CR DROP
    THEN ;
: >inp ( key -- )
    X> inp @                ( D: nybble accum )
    4ROL 0xFFF0 AND OR      ( D: accum' )
    inp ! ;
: >here ( data -- )
    here @ store
    here @1+ ;
: prompt ( -- )
    '>' EMIT BL EMIT ;
: del ( -- )
    cmd @
    DUP BL > IF
        DUP ISHEX IF
            inp @
            4ASR 0x0FFF AND
            DUP IF
                DUP >X
            ELSE
                BL
            THEN cmd !
            inp !
        ELSE
            BL cmd !
        THEN
        '\b' EMIT BL EMIT '\b' EMIT
    THEN DROP ;
: eol ( begin -- end )
    EMIT KEY
    DUP '\r' = SKZ EXIT
    eol ;
: MONITOR
    KEY                     ( D: key )
    DUP ^C = SKZ EXIT       ( abort! )
    DUP '\b' = IF
        DROP DEL
    THEN
    DUP DEL = SKZ del       ( delete previous )
    DUP '/' = SKZ eol       ( comment to EOL )
    DUP EMIT
    DUP '\r' = IF
        '\n' EMIT
    THEN
    cmd @ SWAP              ( D: cmd key )
    ( '<' EMIT OVER X. '.' EMIT DUP X. '>' EMIT )
    DUP BL <= IF
        copy @ IF
            OVER ISHEX IF
                inp @ >here
            THEN
            OVER ']' = IF
                FALSE copy !
            THEN
        ELSE
            OVER ISHEX IF
                inp @ push
            THEN
            OVER '@' = IF
                pop fetch push
            THEN
            OVER '.' = IF
                pop X. CR
            THEN
            OVER '!' = IF
                pop pop SWAP store
            THEN
            OVER 'q' = IF
                pop quad push
            THEN
            OVER '?' = IF
                pop pop SWAP dump
            THEN
            OVER '[' = IF
                pop here !
                TRUE copy !
            THEN
            OVER 'r' = IF
                pop EXECUTE
            THEN
        THEN
        0 inp !             ( clear input accum )
        DUP '\r' = IF
            prompt
        THEN
    THEN
    DUP ISHEX IF
        DUP >inp
    THEN
    DUP DEL = IF
        2DROP
    ELSE
        NIP cmd !           ( key -> cmd )
    THEN
    MONITOR ;

: ECHO
    KEY DUP
    ( EMIT ) X. CR
    DUP '\r' = IF
        '\n' EMIT
    THEN
    ^C = SKZ EXIT           ( abort! )
    ECHO ;

( WARNING! if BOOT returns we PANIC! )
: BOOT
    ECHO prompt MONITOR ;
